netket/optimizer/sr/sr_treemv_logic.py

```diff
@@ -0,0 +1,80 @@
+# Copyright 2021 The NetKet Authors - All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import jax
+import jax.flatten_util
+import jax.numpy as jnp
+
+import numpy as np
+from functools import partial
+
+from netket.stats import sum_inplace, subtract_mean
+from netket.utils import n_nodes
+
+from .sr_onthefly_logic import tree_cast, tree_conj, tree_axpy
+
+
+# TODO cheapest way to calculate the gradients?
+
+
+@partial(jax.vmap, in_axes=(None, None, 0))
+def perex_grads(forward_fn, params, samples):
+    def f(p, x):
+        return forward_fn(p, jnp.expand_dims(x, 0))[0]
+
+    y, vjp_fun = jax.vjp(f, params, samples)
+    res, _ = vjp_fun(np.ones((), dtype=jnp.result_type(y)))
```

Comment:
I don't think this handles R→C functions correctly, because only changes in the real part of the output are backpropagated, so you end up with the real part of the true gradients. This is explained (a bit abstrusely) in the Autodiff cookbook, here is a minimal demonstration:
```
def f(x):
    return x[0] + 1j*x[1]
y, vjp_f = jax.vjp(f, jnp.array([1.0,2.0]))
vjp_f(1.0+0.0j)

>>> (DeviceArray([1., 0.], dtype=float32),)
```
This is not an issue for the lazy solver, because you start with a `jvp` that works fine for R→C functions and apply `vjp` to a complex vector which picks out the right mix of real and imaginary components. Here I don't think you can get away with not splitting an R→C function into real and imaginary parts... But it's neat we can treat R→R and holomorphic cases in one go!

Reply:
The alternative would be building the gradient matrix from `jvp` but that requires you do to do a full forward pass for every parameter in the net, which is prohibitively expensive. I think a `vmap`ped `grad` or `vjp` is the only cheap way of getting the gradient matrix

Reply:
Thanks for catching this !

It was also an issue in the centered on-the-fly code in the part where the mean is precomputed. 
Now the only cases still failing are for inhomogeneous parameters.

The non-centered on-the-fly code (which arguably is the most elegant one and also faster than centered) is and was working correctly also with inhomogeneous parameters.